### PR TITLE
Fix issue getting config from fork join pool

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by geoand** [comment]
cc @radcortez

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/QuarkusConfigFactory.java [hunk 1]
```diff
@@ -34,6 +36,17 @@ public SmallRyeConfig getConfigFor(final SmallRyeConfigProviderResolver configPr
     }
 
     public static void setConfig(SmallRyeConfig config) {
+        //This is a hack, but there is not a great deal that can be done about it
+        //it is possible that when running in test/dev mode some code will be run with the system TCCL
+        //e.g. when using the fork join pool
+        //if this happens any use of config will fail, as the system class loader will resolve the
+        //wrong instance of the config classes
+        if (QuarkusConfigFactory.config != null) {
+            ConfigProviderResolver.instance().releaseConfig(QuarkusConfigFactory.config);
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkus/runtime/configuration/QuarkusConfigFactory.java]: I wonder if we should be catching the ISE here

**Comment by radcortez** [comment]
Thanks! I'll check.

**Comment by radcortez** [comment]
Apparently, this does not fix the issue yet :(

You still get a Classloading issue at this point:
https://github.com/quarkusio/quarkus/blob/b67491c1ffe76248d5f01bc3e29be94c2cdd56a4/core/deployment/src/main/java/io/quarkus/deployment/QuarkusAugmentor.java#L150-L151

I did try to apply the same "hack", at it took me a little further, only to be hit again here:
https://github.com/quarkusio/quarkus/blob/914c83ebd04cca2fd668529c896e819798c1332f/core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java#L190

And here:
https://github.com/quarkusio/quarkus/blob/914c83ebd04cca2fd668529c896e819798c1332f/core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java#L196

And kept going in other places... 

I'm wondering if this is the right way to fix it in all these places.

**Comment by stuartwdouglas** [comment]
I am on PTO this week, will look when I get back

**Comment by radcortez** [comment]
Sure. Thanks!

I do have have some other things I've found. It seems this works fine when running with Java 8:

```
[INFO] Running org.eclipse.microprofile.fault.tolerance.tck.RetryTest
Loaded timeout-multiplier from system property: 1.0
2020-02-03 10:55:26,006 INFO  [io.qua.arc.pro.BeanProcessor] (build-1) Found unrecommended usage of private members (use package-private instead) in application beans:
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForClassLevelMaxRetry,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForDelay,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForMaxRetry,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#retryClientWithNoDelayAndJitter
2020-02-03 10:55:27,017 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 0.716s. Listening on: http://0.0.0.0:8081
2020-02-03 10:55:27,017 INFO  [io.quarkus] (main) Profile test activated.
2020-02-03 10:55:27,018 INFO  [io.quarkus] (main) Installed features: [cdi, smallrye-context-propagation, smallrye-fault-tolerance, smallrye-metrics]
delays are: [0, 268, 0, 143, 243, 0, 55, 40, 0, 0, 240, 0, 0, 379, 124, 0, 0, 39, 85, 183, 0, 0, 139, 260, 133, 232, 377, 0]
2020-02-03 10:55:38,164 INFO  [io.quarkus] (main) Quarkus stopped in 0.032s
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.042 s - in org.eclipse.microprofile.fault.tolerance.tck.RetryTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 306, Failures: 0, Errors: 0, Skipped: 0

java -version
openjdk version "1.8.0_232"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)
```

But with Java 11:

```
[INFO] Running org.eclipse.microprofile.fault.tolerance.tck.RetryTest
Loaded timeout-multiplier from system property: 1.0
2020-02-03 11:09:24,926 INFO  [io.qua.arc.pro.BeanProcessor] (build-27) Found unrecommended usage of private members (use package-private instead) in application beans:
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForClassLevelMaxRetry,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForDelay,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#clientForMaxRetry,
	- @Inject field org.eclipse.microprofile.fault.tolerance.tck.RetryTest#retryClientWithNoDelayAndJitter
2020-02-03 11:09:25,938 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 0.734s. Listening on: http://0.0.0.0:8081
2020-02-03 11:09:25,941 INFO  [io.quarkus] (main) Profile test activated.
2020-02-03 11:09:25,941 INFO  [io.quarkus] (main) Installed features: [cdi, smallrye-context-propagation, smallrye-fault-tolerance, smallrye-metrics]
delays are: [242, 103, 0, 0, 0, 285, 170, 156, 0, 0, 0, 0, 0, 0, 8, 0, 173, 0, 37, 0, 227, 276, 402, 0, 0, 188, 56, 0, 0, 259, 0, 0, 231, 0, 240, 0, 0, 0, 0, 0, 0, 102, 0]
2020-02-03 11:09:38,042 INFO  [io.quarkus] (main) Quarkus stopped in 0.033s
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 16.139 s - in org.eclipse.microprofile.fault.tolerance.tck.RetryTest
[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   RetryConditionTest>Arquillian.run:138->testRetryChainExceptionally:315->assertCompleteExceptionally:359 Cause of ExecutionException
Expected: an instance of java.lang.RuntimeException
     but: <java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype> is a java.util.ServiceConfigurationError
[ERROR]   RetryConditionTest>Arquillian.run:138->testRetryChainSuccess:305->assertCompleteOk:371 Unexpected exceptionjava.util.concurrent.ExecutionException: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
[ERROR]   RetryConditionTest>Arquillian.run:138->testRetryParallelExceptionally:325->assertCompleteExceptionally:359 Cause of ExecutionException
Expected: an instance of java.lang.RuntimeException
     but: <java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype> is a java.util.ServiceConfigurationError
[ERROR]   RetryConditionTest>Arquillian.run:138->testRetryParallelSuccess:335->assertCompleteOk:371 Unexpected exceptionjava.util.concurrent.ExecutionException: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
[INFO]
[ERROR] Tests run: 306, Failures: 4, Errors: 0, Skipped: 0

java -version
openjdk version "11.0.5" 2019-10-15
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.5+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.5+10, mixed mode)
```

And we didn't caught that with the CI, since the TCK's run with Java 8.

**Comment by stuartwdouglas** [comment]
I have updated the fix.

**Comment by radcortez** [comment]
Thanks!

I've just gave it a try and it now passes the Fault Tolerance TCK! 🎉 

Still running the rest.

**Comment by stuartwdouglas** [comment]
@gsmet this should be good to go now

**Review by gsmet — APPROVED** [review]

**Pull request merged**